### PR TITLE
Postgis permission error

### PR DESCRIPTION
**Pull request by enekid**
Reformat `Postgis permission denied for relation` error, removing technical information.

- set error message as: `'permission denied for relation'`
- type as `'auth'` if none set before (eg. `'layer'`)
- add info field with offending relation name

### CODE REVIEW

File: lib/cartodb/middleware/error-middleware.js [hunk 1]
```diff
@@ -70,6 +86,14 @@ function populateTimeoutErrors (errors) {
             error.http_status = 429;
         }
 
+        if (isPermissionDeniedForRelationError(error)) {
+            error.info = getOffendingRelationFromError(error);
```
Comment (dgaubert) [inline lib/cartodb/middleware/error-middleware.js]: Is that change agreed with frontend people?

File: lib/cartodb/middleware/error-middleware.js [hunk 2]
```diff
@@ -54,6 +54,22 @@ function isTimeoutError (err) {
...
+    if (err.message) {
+        const relation = offendingRelationRegex.exec(err.message);
+        if (relation && relation.length > 1) {
+            return relation[1];
+        }
+    }
+    return null;
+}
+
 function populateTimeoutErrors (errors) {
```
Comment (dgaubert) [inline lib/cartodb/middleware/error-middleware.js]: We should change the name of this function, now populated two type of errors: timeout and auth.

File: lib/cartodb/middleware/error-middleware.js [hunk 3]
```diff
@@ -54,6 +54,22 @@ function isTimeoutError (err) {
...
 
+function isPermissionDeniedForRelationError(err) {
+    return err.message && err.message.match(/permission denied for relation/i);   
+}
+
+const offendingRelationRegex = /permission denied for relation (\w+)/i;
+
+function getOffendingRelationFromError(err) {
+    if (err.message) {
+        const relation = offendingRelationRegex.exec(err.message);
```
Comment (dgaubert) [inline lib/cartodb/middleware/error-middleware.js]: being a PITA 👿 : it's actually an array of matches so the name should be plural

File: test/support/test-client.js [hunk 4]
```diff
@@ -47,6 +47,12 @@ module.exports.RESPONSE = {
         headers: {
             'Content-Type': 'application/json; charset=utf-8'
         }
+    },
+    GENERAL_AUTH_ERROR: {
```
Comment (dgaubert) [inline test/support/test-client.js]: `AUTH_ERROR` looks better. doesn't it?

File: lib/cartodb/middleware/error-middleware.js (hunk 3 above)
Comment (enekid) [reply lib/cartodb/middleware/error-middleware.js]: Postgres gives only the name of the first relation that rises the error.

For example, with this query, and having no access granted to neither `table1` nor `table2`:

`select *, (SELECT count(1) from table1) as tc1 from table2`

Postgis would return this error:

`Error: Postgis Plugin: ERROR:  permission denied for relation table2 in executeQuery Full sql was: ...`

Comment (dgaubert) [reply lib/cartodb/middleware/error-middleware.js]: Yeah, I know. I'm talking about `regex` API. The `.exec(`method returns an array of matches. And then we access to the very first element of that array `relation[1]` which the thing that sounds a little weird. As a convention, we use plurals to give a name to `arrays`. For instance: `relations[1]` or `const relation = mathes[1]`.

File: test/support/test-client.js (hunk 4 above)
Comment (enekid) [reply test/support/test-client.js]: There're are 2 types of auth errors:
1. `401` API key not found
2. `403` All other auth errors:
    - _api_ access not granted
    - _dataset_ access not granted

I've created 2 types of auth errors:
- `GENERIC_AUTH_ERROR` -> 403
- `APIKEY_NOT_FOUND_AUTH_ERROR` -> 401

File: test/support/test-client.js [hunk 5]
```diff
@@ -47,6 +47,18 @@ module.exports.RESPONSE = {
         headers: {
             'Content-Type': 'application/json; charset=utf-8'
         }
+    },
+    GENERIC_AUTH_ERROR: {
```
Comment (dgaubert) [inline test/support/test-client.js]: why not use `UNAUTHORIZED`?

File: test/support/test-client.js [hunk 6]
```diff
@@ -47,6 +47,18 @@ module.exports.RESPONSE = {
...
+        status: 403,
+        headers: {
+            'Content-Type': 'application/json; charset=utf-8'
+        }
+    },
+    APIKEY_NOT_FOUND_AUTH_ERROR: {
```
Comment (dgaubert) [inline test/support/test-client.js]: `FORBIDDEN` ??

File: test/support/test-client.js [hunk 7]
```diff
@@ -47,6 +47,18 @@ module.exports.RESPONSE = {
         headers: {
             'Content-Type': 'application/json; charset=utf-8'
         }
+    },
+    UNAUTHORIZED_ERROR: {
```
Comment (dgaubert) [inline test/support/test-client.js]: 403 -> `FORBIDDEN_ERROR `

File: test/support/test-client.js [hunk 8]
```diff
@@ -47,6 +47,18 @@ module.exports.RESPONSE = {
...
+        status: 403,
+        headers: {
+            'Content-Type': 'application/json; charset=utf-8'
+        }
+    },
+    FORBIDDEN_ERROR: {
```
Comment (dgaubert) [inline test/support/test-client.js]: 401 -> `UNAUTHORIZED_ERROR`

File: test/support/test-client.js (hunk 7 above)
Comment (enekid) [reply test/support/test-client.js]: ¯\_(ツ)_/¯ Thanks!

File: test/support/test-client.js (hunk 8 above)
Comment (enekid) [reply test/support/test-client.js]: ¯\_(ツ)_/¯ Thanks!

File: lib/cartodb/middleware/error-middleware.js (hunk 1 above)
Comment (dgaubert) [reply lib/cartodb/middleware/error-middleware.js]: if not, I wouldn't add it.

**Comment by dgaubert** [comment]
Shall we merge this PR?

**Comment by enekid** [comment]
I would put this on hold until we revamp the response error system.

**Comment by dgaubert** [comment]
What does it mean `revamp the response error system`? I'm missing something ¿?

**Comment by enekid** [comment]
This PR doesn't make much sense on it's own until we refactor the whole error management system.

**Pull request closed without merging**